Modules/xxlimited.c

```diff
@@ -152,153 +264,109 @@ xx_foo(PyObject *self, PyObject *args)
-    {"new",             xx_new,         METH_VARARGS,
+    {"new",             xx_new,         METH_NOARGS,
         PyDoc_STR("new() -> new Xx object")},
     {NULL,              NULL}           /* sentinel */
 };
 
+
+/* The module itself */
+
 PyDoc_STRVAR(module_doc,
 "This is a template module just for instruction.");
 
 static int
 xx_modexec(PyObject *m)
 {
-    PyObject *o;
-
-    /* Due to cross platform compiler issues the slots must be filled
-     * here. It's required for portability to Windows without requiring
-     * C++. */
-    Null_Type_slots[0].pfunc = &PyBaseObject_Type;
-    Null_Type_slots[1].pfunc = PyType_GenericNew;
-    Str_Type_slots[0].pfunc = &PyUnicode_Type;
-
-    Xxo_Type = PyType_FromSpec(&Xxo_Type_spec);
-    if (Xxo_Type == NULL)
-        goto fail;
-
-    /* Add some symbolic constants to the module */
-    if (ErrorObject == NULL) {
-        ErrorObject = PyErr_NewException("xxlimited.error", NULL, NULL);
-        if (ErrorObject == NULL)
-            goto fail;
+    xx_state *state = PyModule_GetState(m);
+
+    state->Error_Type = PyErr_NewException("xxlimited.Error", NULL, NULL);
+    if (state->Error_Type == NULL) {
+        return -1;
     }
-    Py_INCREF(ErrorObject);
-    PyModule_AddObject(m, "error", ErrorObject);
-
-    /* Add Xxo */
-    o = PyType_FromSpec(&Xxo_Type_spec);
-    if (o == NULL)
-        goto fail;
-    PyModule_AddObject(m, "Xxo", o);
-
-    /* Add Str */
-    o = PyType_FromSpec(&Str_Type_spec);
-    if (o == NULL)
-        goto fail;
-    PyModule_AddObject(m, "Str", o);
-
-    /* Add Null */
-    o = PyType_FromSpec(&Null_Type_spec);
-    if (o == NULL)
-        goto fail;
-    PyModule_AddObject(m, "Null", o);
+    PyModule_AddType(m, (PyTypeObject*)state->Error_Type);
+
+    state->Xxo_Type = PyType_FromModuleAndSpec(m, &Xxo_Type_spec, NULL);
+    if (state->Xxo_Type == NULL) {
+        return -1;
+    }
+    PyModule_AddType(m, (PyTypeObject*)state->Xxo_Type);
+
+    // Add the Str type. It is not needed from C code, so it is only
+    // added to the module dict.
+    // It does not inherit from "object" (PyObject_Type), but from "str"
+    // (PyUnincode_Type). The tuple of bases is constructed manually.
+    // (It would also possible to fill the Py_tp_base slot, but
+    // it would need to be done dynamically: in standard C, `&PyUnicode_Type`
+    // is not a constant expression .)
+    PyObject *bases = PyTuple_Pack(1, (PyObject *)&PyUnicode_Type);
+    if (bases == NULL) {
+        return -1;
+    }
+    PyObject *Str_Type = PyType_FromModuleAndSpec(m, &Str_Type_spec, bases);
+    Py_DECREF(bases);
+    if (Str_Type == NULL) {
+        return -1;
+    }
+    PyModule_AddType(m, (PyTypeObject*)Str_Type);
+    Py_INCREF(Str_Type);
+
     return 0;
- fail:
-    Py_XDECREF(m);
-    return -1;
 }
 
-
 static PyModuleDef_Slot xx_slots[] = {
     {Py_mod_exec, xx_modexec},
     {0, NULL}
 };
 
+static int
+xx_traverse(PyObject *module, visitproc visit, void *arg) {
```

Comment:
PEP 7 suggests:

```suggestion
xx_traverse(PyObject *module, visitproc visit, void *arg)
{
```

if I recall properly.